### ISSUE TITLE
YOLO-Depth: metric per-pixel distance from a single camera

### ISSUE THREAD

**Issue by raimbekovm** [ultralytics/ultralytics#25695]
**YOLO-Depth** brings monocular depth estimation to Ultralytics YOLO: a metric distance in meters for every pixel, recovered from a single RGB image, using the same familiar `ultralytics` Python and CLI workflows.

This is the launch hub for **YOLO26 depth estimation** models, datasets, outputs, and examples.

## What shipped

- **New depth estimation task**: `depth`
- **YOLO26 depth model family**: `yolo26n-depth.pt`, `yolo26s-depth.pt`, `yolo26m-depth.pt`, `yolo26l-depth.pt`, `yolo26x-depth.pt`
- **Dense metric depth outputs** through `result.depth`
- **Unbounded log-depth head** (~0.02–150 m), so one model covers tabletop and highway ranges alike
- **Closed-form scale calibration** through `model.calibrate()`, already baked into the released weights
- **Built-in dataset configs** for NYU Depth V2, KITTI, Hypersim, SUN RGB-D, ARKitScenes, DIODE, TartanAir, and vKITTI2
- **Train, val, predict, and export workflows** through the standard Ultralytics API

## Start here

- Depth estimation task docs: https://docs.ultralytics.com/tasks/depth
- Depth dataset format: https://docs.ultralytics.com/datasets/depth
- Depth8 quick test dataset: https://docs.ultralytics.com/datasets/depth/depth8
- NYU Depth V2 dataset: https://docs.ultralytics.com/datasets/depth/nyu-depth-v2
- Depth results output: https://docs.ultralytics.com/tasks/depth#results-output
- YOLO26 model docs: https://docs.ultralytics.com/models/yolo26
- Main repository: https://github.com/ultralytics/ultralytics
- Roadmap: https://www.ultralytics.com/roadmap

## Quickstart: CLI

```bash
pip install -U ultralytics

# Predict a dense metric depth map
yolo depth predict model=yolo26n-depth.pt source=https://ultralytics.com/images/bus.jpg imgsz=768

# Train on the tiny Depth8 smoke-test dataset
yolo depth train data=depth8.yaml model=yolo26n-depth.pt epochs=10 imgsz=640

# Validate on NYU Depth V2
yolo depth val model=yolo26n-depth.pt data=nyu-depth.yaml imgsz=768

# Export for deployment
yolo export model=yolo26n-depth.pt format=onnx imgsz=768
```

## Quickstart: Python

```python
from ultralytics import YOLO

model = YOLO("yolo26n-depth.pt")

results = model.predict("https://ultralytics.com/images/bus.jpg", imgsz=768)
for result in results:
    depth = result.depth.data.cpu().numpy()  # H x W float32 map, meters
    print(depth.shape)
    print(depth.min(), depth.max())
```

## Training from custom data

Depth datasets pair each RGB image with a `.npy` float32 depth map in meters. The loader derives the depth path by replacing the `images` component with `depth` and swapping the image extension for `.npy`.

```text
dataset/
├── images/{train,val}
└── depth/{train,val}
```

### Dataset YAML

```yaml
path: /path/to/my-depth-dataset
train: images/train
val: images/val
max_depth: 80 # (m) optional, GT beyond this is excluded from val metrics
nc: 1
names:
    0: depth
```

```bash
yolo depth train data=my-depth.yaml model=yolo26n-depth.pt epochs=100 imgsz=640
```

### Fine-tuning recipe

The released weights are already converged, so fine-tune them with AdamW and a low learning rate instead of the from-scratch SGD defaults:

```bash
yolo depth train data=my-depth.yaml model=yolo26s-depth.pt \
  epochs=20 imgsz=640 optimizer=AdamW lr0=1e-4 warmup_bias_lr=1e-4
```

`mosaic`, `mixup`, `cutmix`, and `copy_paste` are disabled for depth — they combine multiple images, which would produce invalid paired depth maps. Every other augmentation argument applies the same warp to the image and its depth map.

### Calibrating the scale

The head separates scene shape from absolute scale. If relative depth already looks right but the meters are off for your camera, refit the scale in seconds — a closed-form two-parameter fit, no gradient training and no change to the network weights:

```python
from ultralytics import YOLO

model = YOLO("yolo26s-depth.pt")
model.calibrate(data="my-depth.yaml")  # needs a labeled split
model.save("yolo26s-depth-calibrated.pt")
```

## Why it matters

Depth estimation answers a question detection and segmentation cannot: how far away is it. Monocular depth answers it with no stereo rig, no LiDAR, and no calibration setup — one camera in, a full metric map out. That fits robotics and drone navigation, driver assistance, AR/VR occlusion, volumetric measurement, and single-image 3D reconstruction.

The models are built for deployment, not just benchmarks. On a Tesla T4 under TensorRT fp16 at ~768 px, YOLO26n-depth runs in 2.73 ms and YOLO26x-depth in 13.57 ms — both faster than Depth Anything V2 Small at 20.99 ms.

Use YOLO-Depth when you need distance per pixel. Use detection, instance segmentation, or semantic segmentation when you need what the pixels are rather than how far they are.

Share feedback here for YOLO-Depth docs, examples, dataset format questions, and workflows that should be easier to discover.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for sharing the YOLO-Depth launch hub with the Ultralytics community 🚀! This is an automated response, and an Ultralytics engineer will also assist soon.

For users getting started, we recommend visiting the [Docs](https://docs.ultralytics.com/), including the [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples. These resources cover common workflows and may answer many frequently asked questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example), including the model, dataset configuration, command or code, environment details, and complete error output.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image and depth-map examples, training logs, validation results, and verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Confirmed in [PR #25754](https://github.com/ultralytics/ultralytics/pull/25754): `DA3MONO-LARGE` provided structure and `DA3METRIC-LARGE` provided scale, combined using a per-image robust affine fit without camera intrinsics or `focal / 300` scaling. The generated ImageNet maps were stored as float16 `.npy`; for your own dataset, store the resulting meter-scaled maps as float32 per the Ultralytics depth format.

**Comment by Rahulbiradar9** [comment]
Hi @raimbekovm, @glenn-jocher, and @JESUSROYETH,

I noticed that spatial-mixing augmentations like `mosaic`, `mixup`, and `copy_paste` are currently disabled because splicing different images together produces invalid boundary mismatches and relative scale distortions.

Are there plans down the roadmap to support **depth-aware** augmentations? For example:
* Depth-aware object copy-paste (where the pasted object's depth is scaled/adjusted dynamically to fit the target scene's depth contour).
* Correct relative scaling of depth maps when composing mosaics.

**Comment by yzfzzz** [comment]
Are there any structural differences between yolo26-depth and yolo26-detect — is it simply that they share the same backbone but have different heads? If my task requires a single image to output both object bounding boxes and a depth map, how can I fuse these two models — can the backbone be reused? Going a step further, is it possible to use the detection boxes so that the depth model doesn't have to predict depth for all pixels in the image, but only outputs depth for the pixels within the target bounding boxes?

**Comment by Y-T-G** [comment]
> Are there any structural differences between yolo26-depth and yolo26-detect — is it simply that they share the same backbone but have different heads?

Yes, they have the same backbone.

> If my task requires a single image to output both object bounding boxes and a depth map, how can I fuse these two models — can the backbone be reused?

You can combine both heads, but it would require modifying Ultralytics to support two heads.

> Going a step further, is it possible to use the detection boxes so that the depth model doesn't have to predict depth for all pixels in the image, but only outputs depth for the pixels within the target bounding boxes?

Unlikely. They are independent. You can crop the depth map with the boxes as postprocessing. But the depth prediction would be for the whole image before postprocessing.

**Comment by yzfzzz** [comment]
Is the value of result.depth.data the real-world distance?

**Comment by glenn-jocher** [comment]
Yes. Each value in `result.depth.data` is the model’s estimated camera-to-surface distance in meters for that pixel. It is metric depth, but still a monocular estimate; for best absolute accuracy on your camera/domain, calibrate with labeled depth data as described in the [depth scale calibration guide](https://docs.ultralytics.com/tasks/depth#calibrating-the-depth-scale).

**Comment by thi23459112** [comment]
Congratulations on releasing the YOLO Depth weights! 🎉

Would it be possible to use the Depth Map information together with YOLO Detection to generate 3D BBOXes?

Here are some related implementations for reference:
https://www.youtube.com/watch?v=wAKmKsZ9PSw
https://github.com/niconielsen32/YOLO-3D

It would be great if this could eventually be integrated into Ultralytics to support 3D BBOX generation directly.

**Comment by Y-T-G** [comment]
@thi23459112 YOLO 3D is on our roadmap

You can check https://ultralytics.com/roadmap

**Comment by thi23459112** [comment]
> [@thi23459112](https://github.com/thi23459112) YOLO 3D is on our roadmap
> 
> You can check https://ultralytics.com/roadmap

I know that 3D BBOX is already on the roadmap, but as I understand it, the current approach mainly relies on 3D annotated data for training. From our perspective, the cost of generating 3D annotations is simply too high, so we started looking for a more practical alternative.

That’s why I suggested generating 3D BBOX based on 2D BBOX + Depth, using the two corresponding weights. Since YOLO26 already supports both of these components, I think combining them could potentially be the most practical solution for our use case.

**Comment by Y-T-G** [comment]
> I know that 3D BBOX is already on the roadmap, but as I understand it, the current approach mainly relies on 3D annotated data for training. From our perspective, the cost of generating 3D annotations is simply too high, so we started looking for a more practical alternative.
> 
> That’s why I suggested generating 3D BBOX based on 2D BBOX + Depth, using the two corresponding weights. Since YOLO26 already supports both of these components, I think combining them could potentially be the most practical solution for our use case.
> 

What would be the training target in this case? Without 3D annotations, how would the model evaluate whether it's correct in the output?

Or do you mean a general inference only prompt-based depth estimation model that can take in 2D boxes to produce 3D output?

**Comment by thi23459112** [comment]
> it's correct in the output?
> 
> Or do you mean a general inference only prompt-based depth estimation model that can take in 2D boxes to

Exactly. What I’m looking for is a solution that can leverage these two models to produce 3D BBOX results, similar to the effect shown in the YouTube link I shared above.

In my application scenario, I need to use the actual trajectory of objects to make decisions. If I only use the 2D BBOX center point or the bottom-center point to represent the object’s position, the trajectory can become significantly distorted.

The reason is that these 2D reference points do not necessarily represent the object’s actual center in 3D space. That is why I specifically need 3D BBOX information.

However, the annotation cost for 3D BBOX is significantly higher. So I’m wondering whether Ultralytics could potentially provide a branch or an integrated approach that combines these two models, allowing the existing 2D detection results to be extended to 3D BBOX estimation.

The goal would be to achieve more accurate object trajectories while significantly reducing the cost and complexity of 3D annotation and training.


https://github.com/allenai/WildDet3D
this is a SOTA 3D detection.

**Comment by glenn-jocher** [comment]
Yes, an inference-only pipeline is feasible: run detection/segmentation and depth, back-project the object’s depth pixels using calibrated camera intrinsics, reject outliers, then fit a 3D centroid or cuboid and track it over time. Instance masks will generally produce better geometry than rectangular boxes. This can improve trajectories without 3D labels, but monocular depth cannot reliably recover hidden dimensions or orientation, so it is not equivalent to supervised 3D detection. Ultralytics does not currently provide this combined 3D-box pipeline natively, but this is useful feature feedback for a future integration.
